tests/integration/pilot/trafficshifting/helper.go

```diff
@@ -0,0 +1,214 @@
+)
+
+//	Virtual service topology
+//
+//						 a
+//						|-------|
+//						| Host0 |
+//						|-------|
+//							|
+//							|
+//							|
+//		-------------------------------------
+//		|weight1	|weight2	|weight3	|weight4
+//		|a			|b			|c			|d
+//	|-------|	|-------|	|-------|	|-------|
+//	| Host0 |	| Host1	|	| Host2 |	| Host3 |
+//	|-------|	|-------|	|-------|	|-------|
+//
+//
+
+const (
+	testDuration = 10 * time.Second
+
+	errorBand = 10.0
+
+	virtualService = `
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+ name: {{.Name}}
+ namespace: {{.Namespace}}
+spec:
+ hosts:
+ - {{.Host0}}
+ http:
+ - route:
+   - destination:
+       host: {{.Host0}}
+     weight: {{.Weight0}}
+   - destination:
+       host: {{.Host1}}
+     weight: {{.Weight1}}
+   - destination:
+       host: {{.Host2}}
+     weight: {{.Weight2}}
+   - destination:
+       host: {{.Host3}}
+     weight: {{.Weight3}}
+`
+)
+
+var (
+	hosts = []string{"a", "b", "c", "d"}
+)
+
+type VirtualServiceConfig struct {
+	Name      string
+	Host0     string
+	Host1     string
+	Host2     string
+	Host3     string
+	Namespace string
+	Weight0   int32
+	Weight1   int32
+	Weight2   int32
+	Weight3   int32
+}
+
+// RunTrafficShiftingTest configs virtual service and runs traffic shifting test.
+func RunTrafficShiftingTest(t *testing.T, weight []int32) {
+	if len(weight) == 0 || len(weight) > 4 {
+		t.Error("Input parameter is invalid. The length of weight is in [1, 4].")
+	}
+
+	var sum int32
+	for _, v := range weight {
+		if v < 0 || 100 < v {
+			t.Error("Input parameter is invalid. The weight should be in [0, 100].")
+		}
+		sum += v
+	}
+
+	if sum != 100 {
+		t.Error("Input parameter is invalid. The sum of the weight should be 100.0.")
+	}
+
+	weight = append(weight, make([]int32, 4-len(weight))...)
+
+	framework.
+		NewTest(t).
+		RequiresEnvironment(environment.Kube).
+		Run(func(ctx framework.TestContext) {
+			g := galley.NewOrFail(t, ctx, galley.Config{})
+			p := pilot.NewOrFail(t, ctx, pilot.Config{Galley: g})
+
+			instance := apps.NewOrFail(t, ctx, apps.Config{Pilot: p, Galley: g})
+
+			vsc := VirtualServiceConfig{
+				"traffic-shifting-rule",
+				hosts[0],
+				hosts[1],
+				hosts[2],
+				hosts[3],
+				instance.Namespace().Name(),
+				weight[0],
+				weight[1],
+				weight[2],
+				weight[3],
+			}
+
+			tmpl, _ := template.New("VirtualServiceConfig").Parse(virtualService)
+			var buf bytes.Buffer
+			tmpl.Execute(&buf, vsc)
+
+			g.ApplyConfigOrFail(t, instance.Namespace(), buf.String())
+			time.Sleep(10 * time.Second)
+
+			from := instance.GetAppOrFail(hosts[0], t).(apps.KubeApp)
+
+			sendTraffic(t, testDuration, from, hosts[0], weight)
+		})
+}
+
+func sendTraffic(t *testing.T, duration time.Duration, from apps.KubeApp, to string, weight []int32) {
+	timeout := time.After(duration)
+	totalRequests := 0
+	hostnameHitCount := map[string]int{}
+	errorFailures := map[string]int{}
+
+	for {
+		select {
+		case <-timeout:
```

Comment:
This has a hard 10 second deadline before pass/fail detection. Added with 10 second delay above, that is a hard 20 second limit on tests.

Would it make sense to  create a few go-routines and just keep sending requests? After an initial batch size is reached (100?) then you can start periodically checking weight distribution. 

Would such a scheme reduce the runtime duration of the test? (Presumably a passing case should be able to hit the weight ratio in a few seconds, including the initial batch).

Reply:
I ran test locally, the source app sends ~145 requests within 10s so if we start checking weight distribution after the number of request reaches 100, the runtime duration of the test should be reduced. 

I am a little bit confused by 
> After an initial batch size is reached (100?) then you can start periodically checking weight distribution.

When should I stop if I periodically check weight distribution after the number reaches the batch size? If the distribution never hits the weight ratio, the runtime duration will be long due to lots of time is wasted for periodical check.

Maybe we should check immediately after the request number reaches the batch size. If the distribution doesn't hit the expected weight ratio, we fail the test.

Reply:
I ran test locally, the source app sends ~145 requests within 10s so if we start checking weight distribution after the number of request reaches 100, the runtime duration of the test should be reduced. 

I am a little bit confused by 
> After an initial batch size is reached (100?) then you can start periodically checking weight distribution.

When should I stop if I periodically check weight distribution after the number reaches the batch size? If the distribution never hits the weight ratio, the runtime duration will be long due to lots of time is wasted for periodical check.

Maybe we should check immediately after the request number reaches the batch size. If the distribution doesn't hit the expected weight ratio, we fail the test.

Reply:
145 requests probably due to everything being single threaded. If you run the requests on multiple go routines, it should be able to scale much faster I think. You should be able to reach thousands, potentially.

There should still be a timeout, but it should be for failing the test if we cannot get the right distribution in time. However, getting the distribution right should (ideally) happen much earlier than the time out.

Reply:
Make the change in `helper.go`